### PR TITLE
Machine image updates

### PR DESCRIPTION
**Pull request by upodroid**
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/7019
Fixes: https://github.com/hashicorp/terraform-provider-google/issues/7055

I noticed something else. There is a new CMEK field called kmsKeyServiceAccount which allows you to set the Service account to use with KMS. All the resources that use CMEK need to be updated with this field, seen it both v1 and beta apis.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added storage_locations & cmek fields to `google_compute_machine_image` resource
```
```release-note:note
compute: updated `google_compute_machine_image` resource to complete once the Image is ready.
```
```release-note:new-resource
`google_compute_instance_from_machine_image`
```
```release-note:new-resource
`google_compute_machine_image_iam_member`
```
```release-note:new-resource
`google_compute_machine_image_iam_binding`
```
```release-note:new-resource
`google_compute_machine_image_iam_policy`
```

### CODE REVIEW

**Comment by upodroid** [comment]
IAM Tests are passing.

```
REDACTED  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  26⬇  6✎  3+  $    make testacc TEST=./google-beta TESTARGS='-run=TestAccComputeMachineImageIam'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccComputeMachineImageIam -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccComputeMachineImageIamBindingGenerated
=== PAUSE TestAccComputeMachineImageIamBindingGenerated
=== RUN   TestAccComputeMachineImageIamMemberGenerated
=== PAUSE TestAccComputeMachineImageIamMemberGenerated
=== RUN   TestAccComputeMachineImageIamPolicyGenerated
=== PAUSE TestAccComputeMachineImageIamPolicyGenerated
=== RUN   TestAccComputeMachineImageIamBindingGenerated_withCondition
=== PAUSE TestAccComputeMachineImageIamBindingGenerated_withCondition
=== RUN   TestAccComputeMachineImageIamBindingGenerated_withAndWithoutCondition
=== PAUSE TestAccComputeMachineImageIamBindingGenerated_withAndWithoutCondition
=== RUN   TestAccComputeMachineImageIamMemberGenerated_withCondition
=== PAUSE TestAccComputeMachineImageIamMemberGenerated_withCondition
=== RUN   TestAccComputeMachineImageIamMemberGenerated_withAndWithoutCondition
=== PAUSE TestAccComputeMachineImageIamMemberGenerated_withAndWithoutCondition
=== RUN   TestAccComputeMachineImageIamPolicyGenerated_withCondition
=== PAUSE TestAccComputeMachineImageIamPolicyGenerated_withCondition
=== CONT  TestAccComputeMachineImageIamBindingGenerated
=== CONT  TestAccComputeMachineImageIamBindingGenerated_withAndWithoutCondition
=== CONT  TestAccComputeMachineImageIamMemberGenerated_withCondition
=== CONT  TestAccComputeMachineImageIamMemberGenerated_withAndWithoutCondition
=== CONT  TestAccComputeMachineImageIamPolicyGenerated_withCondition
=== CONT  TestAccComputeMachineImageIamPolicyGenerated
=== CONT  TestAccComputeMachineImageIamBindingGenerated_withCondition
=== CONT  TestAccComputeMachineImageIamMemberGenerated
--- PASS: TestAccComputeMachineImageIamBindingGenerated_withCondition (158.54s)
--- PASS: TestAccComputeMachineImageIamMemberGenerated_withCondition (159.43s)
--- PASS: TestAccComputeMachineImageIamPolicyGenerated (163.24s)
--- PASS: TestAccComputeMachineImageIamBindingGenerated (168.55s)
--- PASS: TestAccComputeMachineImageIamPolicyGenerated_withCondition (170.35s)
--- PASS: TestAccComputeMachineImageIamBindingGenerated_withAndWithoutCondition (171.29s)
--- PASS: TestAccComputeMachineImageIamMemberGenerated_withAndWithoutCondition (171.72s)
--- PASS: TestAccComputeMachineImageIamMemberGenerated (171.72s)
PASS
ok      github.com/hashicorp/terraform-provider-google-beta/google-beta 173.983s
```

**Comment by upodroid** [comment]
In commit c7820aa , the test are failing because the logic to override disks is broken.

```
maha4472  MCW0CDP3YY  ~  go  …  github.com  hashicorp  terraform-provider-google-beta   master  26⬇  10✎  6+  USAGE  $    make testacc TEST=./google-beta TESTARGS='-run=TestAccComputeInstanceFromMachineImage'
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccComputeInstanceFromMachineImage -timeout 240m -ldflags="-X=github.com/hashicorp/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccComputeInstanceFromMachineImage_basic
=== PAUSE TestAccComputeInstanceFromMachineImage_basic
=== RUN   TestAccComputeInstanceFromMachineImage_overrideBootDisk
=== PAUSE TestAccComputeInstanceFromMachineImage_overrideBootDisk
=== RUN   TestAccComputeInstanceFromMachineImage_overrideAttachedDisk
=== PAUSE TestAccComputeInstanceFromMachineImage_overrideAttachedDisk
=== RUN   TestAccComputeInstanceFromMachineImage_overrideScratchDisk
=== PAUSE TestAccComputeInstanceFromMachineImage_overrideScratchDisk
=== RUN   TestAccComputeInstanceFromMachineImage_overrideMetadataDotStartupScript
=== PAUSE TestAccComputeInstanceFromMachineImage_overrideMetadataDotStartupScript
=== CONT  TestAccComputeInstanceFromMachineImage_basic
=== CONT  TestAccComputeInstanceFromMachineImage_overrideScratchDisk
=== CONT  TestAccComputeInstanceFromMachineImage_overrideBootDisk
=== CONT  TestAccComputeInstanceFromMachineImage_overrideAttachedDisk
=== CONT  TestAccComputeInstanceFromMachineImage_overrideMetadataDotStartupScript
    TestAccComputeInstanceFromMachineImage_overrideBootDisk: testing.go:674: Step 0 error: Check failed: Check 3/3 error: google_compute_instance_from_machine_image.foobar: Attribute 'boot_disk.0.name' didn't match "terraform-test-0dymegdnmb", got ""
    TestAccComputeInstanceFromMachineImage_overrideScratchDisk: testing.go:674: Step 0 error: Check failed: Check 3/3 error: google_compute_instance_from_machine_image.foobar: Attribute 'scratch_disk.0.interface' expected "NVME", got "SCSI"
    TestAccComputeInstanceFromMachineImage_overrideAttachedDisk: testing.go:674: Step 0 error: Check failed: Check 1/3 error: Not found: google_compute_instance_from_machine_image.inst
--- PASS: TestAccComputeInstanceFromMachineImage_basic (287.22s)
--- FAIL: TestAccComputeInstanceFromMachineImage_overrideAttachedDisk (310.96s)
--- FAIL: TestAccComputeInstanceFromMachineImage_overrideBootDisk (332.47s)
--- FAIL: TestAccComputeInstanceFromMachineImage_overrideScratchDisk (389.99s)
--- PASS: TestAccComputeInstanceFromMachineImage_overrideMetadataDotStartupScript (403.27s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-google-beta/google-beta 405.797s
FAIL
make: *** [testacc] Error 1
```

**Comment by upodroid** [comment]
I applied a similar logic in `adjustInstanceFromMachineImageDisks` from the `google_compute_instance_from_instance_template` but i'm getting inconsistent state behaviour. I suspect it is the type changes in that function.

https://gist.github.com/upodroid/1721c4e4ee26930619813a182c1e1be2 Results for the acceptance test.

File: products/compute/api.yaml
Comment (c2thorn) [inline products/compute/api.yaml:7090]:
```suggestion
```

File: third_party/terraform/resources/resource_compute_instance_from_machine_image.go.erb
Comment (c2thorn) [inline third_party/terraform/resources/resource_compute_instance_from_machine_image.go.erb:80]:
```suggestion
```

File: third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown
Comment (c2thorn) [inline third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown:58]: Mind splitting the second sentence here into its own Attribute Reference? Similar to https://github.com/GoogleCloudPlatform/magic-modules/pull/3877/files

File: third_party/terraform/resources/resource_compute_instance_from_machine_image.go.erb
Comment (c2thorn) [inline third_party/terraform/resources/resource_compute_instance_from_machine_image.go.erb:37]:
```suggestion
		// the machine image values will be used if they're unset.
```

File: third_party/terraform/resources/resource_compute_instance_from_machine_image.go.erb
Comment (c2thorn) [inline third_party/terraform/resources/resource_compute_instance_from_machine_image.go.erb:55]:
```suggestion
		// machine image, so remove defaults.
```

File: third_party/terraform/resources/resource_compute_instance_from_machine_image.go.erb
Comment (c2thorn) [reply third_party/terraform/resources/resource_compute_instance_from_machine_image.go.erb:80]: Can just be removed

File: third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown
Comment (c2thorn) [inline third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown:29]:
```suggestion
  name = "instance-from-machine-image"
```

File: third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown
Comment (c2thorn) [inline third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown:50]:
```suggestion
* `source_machine_image` - (Required) Name or self link of a machine
  image to create the instance based on.
```

File: third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown
Comment (c2thorn) [inline third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown:58]:
```suggestion
as a way to override the properties in the machine image. All exported attributes
```

File: third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown
Comment (c2thorn) [inline third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown:28]: We should specify the google-beta provider in the example

File: third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown
Comment (upodroid) [reply third_party/terraform/website/docs/r/compute_instance_from_machine_image.html.markdown]: Fixed

**Comment by upodroid** [comment]
Applied the minor changes, the override logic is still broken.

**Comment by upodroid** [comment]
@c2thorn ^

**Comment by c2thorn** [comment]
Hi @upodroid, sorry about that. I was under the impression that you still wanted to address the override logic. Do you have any ideas for that?

**Comment by upodroid** [comment]
Not really. I'm stuck on making the disk override work correctly.

**Comment by c2thorn** [comment]
Okay, I'll take a look at it!

File: products/compute/api.yaml
Comment (drebes) [inline products/compute/api.yaml:7065]: This description does not match `storageLocations`.

Comment (upodroid) [reply products/compute/api.yaml:7065]: I'll fix it.

**Comment by c2thorn** [comment]
@upodroid I found out internally that there is a Compute bug for overriding disks from machine images. Your logic for that should be correct. Allegedly it should have already rolled out, but I am still running into failures using the latest googleapi go client. I'll follow up internally.

Also, I don't mind adding a commit or two to this PR, but I think I need push access to your fork.

**Comment by upodroid** [comment]
Thanks for the update

**Comment by c2thorn** [comment]
Hmm still getting a 403 trying to add to this PR. @upodroid can you try checking the box on this page stating `Allow edits and access to secrets by maintainers` ?

**Comment by upodroid** [comment]
I can't see it.

I ticked it the first time I opened a PR from my fork.

**Comment by upodroid** [comment]
image

**Comment by upodroid** [comment]
Urgh, the fork belongs to my gh org and not my user

**Comment by upodroid** [comment]
I updated the release notes to remove the iam resources. They were implemented in a separate PR.

**Comment by c2thorn** [comment]
@upodroid The machine image IAM resources? can't seem to find them in any other PR/in TPGB currently

**Comment by c2thorn** [comment]
Also, I am planning to move forward with this PR with the minor changes I've made. I've basically updated the PR to be compatible with general changes to the repo, and then removed the ability to override disks (not yet supported by the API). Testing changes now

**Comment by upodroid** [comment]
IAM was done in https://github.com/GoogleCloudPlatform/magic-modules/pull/3887 . I remember it because the issue was duplicated.

**Comment by upodroid** [comment]
nvm, that was compute image not machine image :D

**Comment by upodroid** [comment]
Sounds great, I just need to fix the typo @drebes pointed out earlier

**Comment by c2thorn** [comment]
@upodroid already taken care of!

**Comment by c2thorn** [comment]
Tests pass! I know it has been a while on this one, thanks for your help and patience. I'll likely merge this in later today if you don't have any concerns.

**Comment by upodroid** [comment]
Thanks, merge it once you are ready.

**Review by c2thorn — APPROVED** [review]
Might as well do it soon.

**Pull request merged**
